MAKE @SMITTYVB LORD CHANCELLOR FOREVER!

mark-i-m opened the pull request:
- If chaosbot crashes within the first 60s of starting, it is restarted by supervisord up to 10 times before supervisord finally gives up. This gives chaosbot a chance to go through at least one poll/merge cycle.

- Every time chaosbot crashes it reverts to an older commit in the git history in the hope that this resolves the issue. To prevent attackers from hiding malicious code in their PRs and getting chaosbot to rollback to them, we only rollback to commits by chaosbot itself (i.e. merges).

- When chaosbot crashes, it posts an issue on github

- Meanwhile, if there is a pull request for it to pull and merge, it does so after checking out master, so chaosbot will never get stuck in the past.

- Rollbacks are disabled unless the git remote called `origin` is `chaosbot/Chaos`

- Throw in a few typo and minor bug fixes, too...

EDIT: updated rollback stuff
EDIT: updated again with recent developements
EDIT: To be clear, this PR has nothing to do with changing the voting system... The title is mostly just to attract attention :laughing:

mark-i-m committed:
Make chaos rollback if it induces failures

mark-i-m committed:
Fix typos and minor bugs with flushing before execl

mark-i-m committed:
Add git to the vagrant file

mark-i-m committed:
Create a dummy chage!

mark-i-m committed:
fixed supervisor conf

mark-i-m committed:
fixed typo

phil-r commented on dev/vagrant/Vagrantfile:
```diff
@@ -14,6 +14,7 @@ Vagrant.configure("2") do |config|
   # Ensure basic python requirements are initialized.
   config.vm.provision "shell", inline: <<-SCRIPT
     apt-get install --assume-yes\
+	git\
```
strange indentation

mark-i-m replied:
Oh,  it looks like tabs vs spaces... will fix

mark-i-m replied:
fixed

PlasmaPower commented on chaos_wrapper.sh:
```diff
@@ -0,0 +1,8 @@
+    git checkout HEAD^
+}
+
+/root/.virtualenvs/chaos/bin/python chaos.py || rollback
```
Maybe only rollback if chaos crashes within 60 seconds?

phil-r replied:
Basically rollback tries only the previous commit (this can be a cool attack vector btw, because previous commit can be in the same PR that crashed it)

PlasmaPower replied:
That would be an amazing attack vector! I didn't think about that. Maybe revert up to and including the last chaosbot commit?

mark-i-m replied:
Yeah, I thought about this a bit because I've accidentally done it to myself a few times :stuck_out_tongue: 

> Maybe revert up to and including the last chaosbot commit

That sounds like a viable solution

mark-i-m committed:
fixed weird indentation

PlasmaPower commented on cron/poll_pull_requests.py:
```diff
@@ -118,10 +119,18 @@ def poll_pull_requests():
         startup_path = join(THIS_DIR, "..", "startup.sh")
+
+        # before we exec, we need to flush i/o buffers so we don't lose logs or voters
+        sys.stdout.flush() 
```
Trailing whitespace - Travis is failing.

mark-i-m committed:
whitespace

amoffat commented on cron/poll_pull_requests.py:
```diff
@@ -118,10 +119,18 @@ def poll_pull_requests():
             json.dump(old_votes, fp)
 
+            # flush all buffers because we might restart, which could cause a crash
+            os.fsync(fp)
```
https://github.com/chaosbot/Chaos/blob/master/memoize/backends.py#L44-L49

might help us here

mark-i-m replied:
Ah, thanks :)

mark-i-m replied:
@amoffat Are you suggesting that I directly use `JSONBackend`? Or that it is already used? Or that I should do something in the same vein?

amoffat replied:
@mark-i-m sorry I made a vague comment and then had to leave :)  Don't worry about it...I saw you were looking at making the json writing a little more resilient to restarts, and it reminded me that we need to make this part write to a backup file, then atomic move to the correct file, like in the code I linked, so if we restart or crash during `json.dumps`, our json file won't be corrupted.

I realize that this is probably outside the scope of the PR so don't worry about it, I can take a look at it :)

mark-i-m replied:
@amoffat Thanks :smile: Perhaps in my next PR.

mark-i-m committed:
Merge branch 'master' of https://github.com/chaosbot/Chaos

PlasmaPower commented on chaos_wrapper.sh:
```diff
@@ -0,0 +1,8 @@
+
+function rollback() {
+    git checkout master
+    git checkout HEAD^
```
Replace both these commands with `git reset --hard $(git log --author chaosbot --format=format:%H | tail -n +2 | head -n 1)`.

PlasmaPower replied:
That doesn't push the changes or anything, the actual fix is left up to us. This just gives us the opportunity to make it.

mark-i-m replied:
Actually, it occurs to me that the use of `git checkout master` is completely nonsensical here because then we will never get past the second-to-last commit :face_with_head_bandage:

PlasmaPower replied:
Yeah, I was wondering if that was intentional...

mark-i-m replied:
So now the question is should we do it always if there is an abnormal exit or only if it crashes in the first 60s, as you suggested [above](https://github.com/chaosbot/Chaos/pull/318#discussion_r118827775)...

mark-i-m committed:
better chaos_wrapper.sh is less conducive to attack

mark-i-m committed:
Fix bug

mark-i-m committed:
Reread supervisor conf after rollback

mark-i-m committed:
Handle the case where there is a failure after 60s

mark-i-m committed:
Fixed minor bug

mark-i-m committed:
Apparently you need to restart too

mark-i-m committed:
Still trying to make supervisor behave

xyproto commented on etc/chaos_supervisor.conf:
```diff
@@ -1,5 +1,5 @@
 [program:chaos]
-command=/root/.virtualenvs/chaos/bin/python chaos.py
+command=bash chaos_wrapper.sh
```
`./chaos_wrapper.sh` should also work, provided that `chaos_wrapper.sh` has the executable bit set, since `#!/bin/bash` is at the top of `chaos_wrapper.sh`.

mark-i-m replied:
Surprisingly, it doesn't. That's actually what I had originally, and I had to fiddle with it to get it to work. I suspect that supervisord is actually `exec`-ing the command or something in a way that doesn't know about interpreters...

xyproto replied:
That's a surprise. Perhaps it would work if the full path to the script was given?

mark-i-m replied:
Yeah, but I don't know what the full path would be on production. Is there a particular reason why that would be better?

xyproto replied:
nah, it's not that important, only marginally cleaner, IMO

phil-r commented on chaos_wrapper.sh:
```diff
@@ -0,0 +1,31 @@
+
+function rollback() {
+    # avoid commits from the same faulty PR
+    rb_commit=$(git log --author chaosbot --format=format:%H | tail -n +2 | head -n 1)
```
It looks like commits with author `chaosbot` happen when @amoffat manually fixes problems, this should be changed to `Chaos`

mark-i-m replied:
Ah, that makes sense. I was wondering why it was rolling back so many commits.

mark-i-m replied:
Fixed

mark-i-m committed:
Use the correct author for rollback

mark-i-m committed:
merging PR #1: Create a dummy change!

https://github.com/mark-i-m/Chaos/pull/1: Create a dummy change!

Description:


:ok_woman: PR passed with a vote of 1 for and 0 against, with a weighted total             of 1.0 and a threshold of 1.0.

Vote record:
@mark-i-m: 1

mark-i-m committed:
Revert "Create a dummy change!"

mark-i-m committed:
Merge pull request #2 from mark-i-m/revert-1-dummy

Sorry, I didn't mean to upvote this...

phil-r commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# ChaosBot
```
why? =(

mark-i-m replied:
Did I mess it up? I had a created a fake PR for testing that chained the title in the README to `ChaosBoat`. Except my instance of chaosbot actually merged it, so I reverted it.

phil-r commented on chaos_wrapper.sh:
```diff
@@ -0,0 +1,31 @@
+
+function rollback() {
+    # avoid commits from the same faulty PR
+    rb_commit=$(git log --author Chaos --format=format:%H | tail -n +2 | head -n 1)
```
After thinking a bit more about it I realised that it's still not bulletproof 
"attacker" can just make a commit like this: `git commit --author "Chaos"` 

I'm afraid we can't solely rely on git for this ;(

PlasmaPower replied:
Yeah, I know that, but unless we want chaosbot GPG signing commits there's no other way. In addition, we can spot PRs with commits with an author of Chaos.

mark-i-m replied:
TBH, I considered creating such a commit (you can imagine what I might have put in it) :stuck_out_tongue:

mark-i-m replied:
Security wise, there is a much deeper problem: as soon as somebody starts their own cron job, they can basically turn chaosbot into a... well, bot (listening to a C&C server). This is not very hard, as we saw with #77... You just need to hide a commit that contains some nefarious code behind a commit that reverts it. That way, your code is in the repo. If at any point you can manage to get the production server to git checkout and run your own copy, then you have backdoored chaosbot, kind of like [Ken Thompson's famous talk](http://www-inst.eecs.berkeley.edu/~cs161/fa08/papers/p761-thompson.pdf).

phil-r replied:
@PlasmaPower yeah it will look a bit stange, but it's just my avatar 2 times: [example PR](https://github.com/chaosbot/Chaos/pull/326)
if you'll run `git log --author Chaos` my commit will be there

PlasmaPower replied:
Well we can look for the double avatar.

PlasmaPower replied:
@mark-i-m yeah I considered adding a SSH key for myself when I had dictatorship

mark-i-m committed:
Now that we are squashing...

mark-i-m committed:
Merge branch 'master' of github.com:mark-i-m/Chaos

mark-i-m committed:
Merge branch 'master' of https://github.com/chaosbot/Chaos

mark-i-m committed:
Merge branch 'master' of https://github.com/chaosbot/Chaos

mark-i-m committed:
Post to github on crashes

mark-i-m committed:
Fixed a couple of bugs

mark-i-m committed:
lints

mark-i-m committed:
Only rollback in production

PlasmaPower reviewed (approved):
Looks good to me!
